### ISSUE TITLE
Restore 2 Support

### ISSUE THREAD

**Issue by cameronalberg** [dgreif/homebridge-hatch-baby-rest#104]
With the [new Restore 2](https://www.hatch.co/restore-2?utm_source=google_search&utm_medium=paid_search&gclid=CjwKCAjwiOCgBhAgEiwAjv5whHFAku3f-7VtdzTysLfRis-YHD16UwVMkI_ACrho2tBD6QeU_xzYCBoC2xAQAvD_BwE), getting the following message in homebridge: 
`Unsupported Product Found: restoreV4`

Any plans to support this model? I tried tinkering with the plugin to recognize it, without much success. Happy to help, but unsure on how to pick up the MQTT payloads being sent through the app.

**Comment by dgreif** [comment]
Thanks for offering to help! There is a [comment on another issue](https://github.com/dgreif/homebridge-hatch-baby-rest/issues/99#issuecomment-1442557545) which covers roughly how to gather the data.

**Comment by simplytoast1** [comment]
Has anyone does this?  If not I have no issue buying one to get you the information you need.

**Comment by dgreif** [comment]
@simplytoast1 no, I haven't seen the data for the `restoreV4` posted by anyone yet. I don't have a lot of free time at the moment, so no promises on if/when I would get around to implementing it even if the data is posted 😞

**Comment by dgreif** [comment]
@mpatfield @simplytoast1 I've added a new `debug` option which will log all the device info/state which is what I need in order to add support. Brief instructions can now be found [in the README](https://github.com/dgreif/homebridge-hatch-baby-rest/tree/main/packages/homebridge-hatch-baby-rest#device-discovery). Would one of you mind running the latest version with `"debug": true` and posting the info/state here?

**Comment by mpatfield** [comment]
Thanks @dgreif! Unfortunately, I'm not seeing any additional info in the logs.

`[10/02/2024, 08:13:56] Homebridge v1.7.0 (HAP v0.11.1) (Homebridge) is running on port 51488.`
`[10/02/2024, 08:14:08] [HatchBabyRest] Unsupported Product Found: restoreV4`
`[10/02/2024, 08:14:08] [HatchBabyRest] Configuring Hatch Devices:`
`[10/02/2024, 08:14:09] [HatchBabyRest]   No supported devices found`

To confirm, I updated to `v4.3.0` and added `"debug": true` to the config

`{`
` "debug": true,`
` "email": "[redacted]",`
` "password": "[redacted]",`
` "platform": "HatchBabyRest"`
`}`

I also turned on debug mode for Homebridge `-D` thinking that might help but I'm still not seeing anything. It should be showing up in the `homebridge.log` right? Or is it buried in some other log?

**Comment by dgreif** [comment]
@mpatfield that's really odd. I just double checked and it looks like I'm checking the config properly. Maybe the update failed? Did you for sure restart homebridge?

**Comment by mpatfield** [comment]
Pretty sure the plugin update worked because I’m not seeing ‘alexa’ as an unsupported device anymore. I just restarted again to make sure and still no dice. 😕

Sorry for being so high maintenance. I really appreciate all the effort you are putting in especially since there’s no direct benefit to you.

Let me know if there’s anything else I can try or we can wait to see if @simplytoast1 has more luck.

**Comment by mpatfield** [comment]
@dgreif besides adding `"debug": true` to the plugin config, I tried running homebridge with the `-D` option and `DEBUG=*` environment variable. Is there anything else you can think of I would need to do in order to see the debug output you're looking for?

**Comment by vdiaz311** [comment]
Just chiming in to follow the thread. I also tried the debug instructions to no dice.

**Comment by dgreif** [comment]
Wow, sorry for seriously gas lighting y'all! @mpatfield dropped a [comment](https://github.com/dgreif/homebridge-hatch-baby-rest/commit/53769b79e06f2608d262dd9fbea1ff11ca861e96#r138847177) which got me headed in the right direction. Give it another try in `4.3.1` 🙏 🤞

**Comment by vdiaz311** [comment]
> @mpatfield @simplytoast1 I've added a new `debug` option which will log all the device info/state which is what I need in order to add support. Brief instructions can now be found [in the README](https://github.com/dgreif/homebridge-hatch-baby-rest/tree/main/packages/homebridge-hatch-baby-rest#device-discovery). Would one of you mind running the latest version with `"debug": true` and posting the info/state here?

Is this what you're looking for? 

Info:

id": "***",
  "createDate": "2023-12-16 21:23:53",
  "updateDate": "2023-12-16 21:23:53",
  "macAddress": "***",
  "owner": true,
  "name": "Emily’s Restore",
  "hardwareVersion": "10.0.2",
  "product": "restoreV4",
  "email": "***",
  "thingName": "***",
  "memberId": "***"

State:

"env": "prod",
  "alarmsDisabled": false,
  "nightlightOn": false,
  "nightlightIntensity": 32780,
  "toddlerLockOn": false,
  "snoozeDuration": 540,
  "knockThreshold": 20,
  "knockDuration": 0,
  "activeTap": true,
  "knockAxis": 1,
  "current": {
    "srId": 0,
    "playing": "none",
    "step": 0,
    "color": {
      "id": 0,
      "r": 0,
      "g": 0,
      "b": 0,
      "w": 0,
      "i": 0,
      "duration": 0,
      "until": "indefinite"
    },
    "sound": {
      "id": 19998,
      "v": 16383,
      "mute": false,
      "url": "https://images.hatchbaby.com/windDowns/Rest.mp3",
      "duration": 0,
      "until": "indefinite"
    }
  },
  "dataVersion": "20240306110438089",
  "sleepScene": {
    "srId": 0,
    "enabled": false
  },
  "timer": {
    "s": "2018-09-17 16:35:28",
    "d": 0
  },
  "streaming": {
    "status": "none"
  },
  "timezone": "America/Los_Angeles",
  "rF": {
    "v": "10.1.448",
    "i": true,
    "u": "https://firmware.hatchbaby.com/prod/restoreV4/restoreV4_DFU_10_1_448.json?f=g"
  },
  "deviceInfo": {
    "f": "10.1.448",
    "fR": 0,
    "hwVersion": "10.0.2",
    "powerStatus": 0
  },
  "clock": {
    "i": 429562265,
    "turnOffAt": "22:00:00",
    "turnOnAt": "07:00:00",
    "turnDimAt": "22:00:00",
    "turnBrightAt": "07:00:00",
    "flags": 16,
    "turnOffMode": "always"
  },
  "toddlerLock": {
    "turnOffAt": "00:00:00",
    "turnOnAt": "00:00:00",
    "turnOnMode": "never"
  },
  "lucky": 0,
  "LDR": "UpgReset",
  "LWTP": false,
  "debug": 0,
  "hwDebugFlags": 0,
  "logging": 37,
  "owned": true,
  "lastReset": "PowerOn_OtherWDT",
  "ota": {
    "status": "none",
    "downloadProgress": 0,
    "installProgress": 0
  },
  "touch": {
    "flags": 0,
    "poll_rate_hz": 0,
    "refire_delay_ms": 0,
    "refire_rate_hz": 0,
    "step_size": 0
  },
  "REX": {
    "lock": 2864466494,
    "key": 0,
    "command": "none"
  },
  "connected": true,
  "rssi": -14,
  "snooze": {
    "active": false,
    "startTime": ""

**Comment by vdiaz311** [comment]
> Wow, sorry for seriously gas lighting y'all! @mpatfield dropped a [comment](https://github.com/dgreif/homebridge-hatch-baby-rest/commit/53769b79e06f2608d262dd9fbea1ff11ca861e96#r138847177) which got me headed in the right direction. Give it another try in `4.3.1` 🙏 🤞

Good work, man. See my comment above and let me know if that's what you were looking for.

**Comment by dgreif** [comment]
That's exactly what I needed @vdiaz311, thank you! Personal life is very hectic at the moment, so I can't commit to actually implementing Restore 2 support on any specific timeline, but this is the info I need to do it 🎉

**Comment by vdiaz311** [comment]
> That's exactly what I needed @vdiaz311, thank you! Personal life is very hectic at the moment, so I can't commit to actually implementing Restore 2 support on any specific timeline, but this is the info I need to do it 🎉

No worries on the timeline. The fact you're even looking into this at all is great.

**Comment by eoddc5** [comment]
no worries on the timeline, but wanted to check if theres any progress or updates on this feature since the March 12 messages?

**Comment by heyitsericc** [comment]
Hi, another Hatch 2 owner here with the same situation. 

Unsupported Product Found: restoreV4

I added "debug": true and am getting the same results as those above.

Still quite new to Homebridge, wondering if there will be any updates to this?

Thank you in advance for your help.

**Comment by noahthomas03** [comment]
@dgreif any update on getting support for Restore 2? I just bought one thinking this plugin supported it, but I apparently was wrong. I hope there's a chance you're still willing to add support for it. Thanks!

**Comment by dgreif** [comment]
Thank you all for being patient. This is really low priority for me, so not sure when I'll get around to it. If someone else wants to take a stab at writing a PR for it, I'd be happy to review it

**Comment by chase9** [comment]
I have initial support for the Restore V2 if anyone wants to help test!

https://github.com/chase9/homebridge-hatch-baby-rest

So far it only seems to turn the device off. I think we'll need to play with the parameters a bit to find out what triggers the routine.

**Comment by djp-18** [comment]
Any update on this? I'd really like to get this to work

**Comment by chase9** [comment]
I haven't been working on it as I found out the wrong settings can make it so that support needs to reset your restore. Since I have a newborn now, I'm unwilling to risk that (at least for now 😄 )

**Comment by djp-18** [comment]
Interesting.. I spent about 4-5 hours troubleshooting with chatgpt and it is working pretty reliably with my restore2 so far. Do you know where the wrong settings are? I'll get around to it eventually

**Comment by Tmatz27** [comment]
Im also following with the same issue. Looking to get this working but also understand the situation.

**Comment by djp-18** [comment]
I’ve got a working plugin with my restore v2… I’d be willing to share with @dgreif and @chase9 so that everyone can benefit. Thanks!

![Image](https://github.com/user-attachments/assets/12f97608-2b76-479c-aaf0-f16bb6e2f9c1)

**Comment by chase9** [comment]
Right, your credentials should only be in your code if you put them in there. Thanks!

**Comment by djp-18** [comment]
I've got the zip ready to go, could you message me your email? I don't believe you have one on your profile. Thanks!

**Comment by chase9** [comment]
Maybe this would work best:

https://nextcloud.chaselau.me/s/FyBDRMWTKjNtrDT

This will let you upload your file to my nextcloud instance so I don't have to type my email address on the internet :)

**Comment by djp-18** [comment]
Appears the link is no longer valid - my apologies. I traveled over the weekend and did not save my .zip to any cloud storage. I'm ready to upload now.

**Comment by chase9** [comment]
Alright, I set this one for a week :)

https://nextcloud.chaselau.me/s/8SPGJF3Ejy5ykot

**Comment by chase9** [comment]
@djp-18 Just now getting to this. It looks like you uploaded several thousand files and it doesn't look like any of them were from this project? Do you want to upload just a ZIP of your changed files somewhere and I can commit them?

**Comment by benedicttobias** [comment]
I also could not risk the corrupted config, unfortunately. I didn't realized that [RestoreState](https://github.com/dgreif/homebridge-hatch-baby-rest/blob/780c70b0674f4298f38656788f711ee173d7be03/packages/shared/hatch-sleep-types.ts#L322) interface is play a role of writing the json... and of course it is different than Restore 2...

I guess we have to create a new "RestoreV4State" 

Now my config out there has the extra json below 🤣 
```json
"content": {
    "playing": "routine",
    "paused": false,
    "offset": 0,
    "step": 1
  }
```

I attempt to get two state of the config:

This is the state of 'do nothing'
```json
{
  "env": "prod",
  "alarmsDisabled": false,
  "nightlightOn": false,
  "nightlightIntensity": 32780,
  "toddlerLockOn": false,
  "snoozeDuration": 540,
  "knockThreshold": 20,
  "knockDuration": 0,
  "activeTap": true,
  "knockAxis": 1,
  "current": {
    "srId": 0,
    "playing": "none",
    "step": 0,
    "color": {
      "id": 0,
      "r": 0,
      "g": 0,
      "b": 0,
      "w": 0,
      "i": 0,
      "duration": 0,
      "until": "indefinite"
    },
    "sound": {
      "id": 19998,
      "v": 16384,
      "mute": false,
      "url": "https://images.hatchbaby.com/windDowns/Rest.mp3",
      "duration": 0,
      "until": "indefinite"
    },
    "paused": false
  },
  "dataVersion": "20250823151851237",
  "sleepScene": {
    "srId": 0,
    "enabled": false
  },
  "timer": {
    "s": "2024-05-12 00:26:40",
    "d": 0
  },
  "streaming": {
    "status": "none"
  },
  "timezone": "America/New_York",
  "rF": {
    "v": "10.1.575",
    "i": true,
    "u": "https://firmware.hatchbaby.com/prod/restoreV4/restoreV4_DFU_10_1_575.json?f=g"
  },
  "deviceInfo": {
    "f": "10.1.575",
    "fR": 0,
    "hwVersion": "10.0.2",
    "powerStatus": 0,
    "sdCardVersionInfo": {
      "releaseDate": "unknown",
      "hashType": "md5",
      "hashCode": "0xe4e5d909ca46a1633e42ffe7b99e12ae"
    }
  },
  "clock": {
    "i": 429531702,
    "turnOffAt": "23:59:00",
    "turnOnAt": "07:00:00",
    "turnDimAt": "23:59:00",
    "turnBrightAt": "07:00:00",
    "flags": 34816,
    "turnOffMode": "custom"
  },
  "toddlerLock": {
    "turnOffAt": "00:00:00",
    "turnOnAt": "00:00:00",
    "turnOnMode": "never"
  },
  "lucky": 0,
  "LDR": "UNPOWERED",
  "LWTP": false,
  "debug": 0,
  "hwDebugFlags": 0,
  "logging": 37,
  "owned": true,
  "lastReset": "PowerOn",
  "ota": {
    "status": "none",
    "downloadProgress": 0,
    "installProgress": 0
  },
  "touch": {
    "flags": 0,
    "poll_rate_hz": 0,
    "refire_delay_ms": 0,
    "refire_rate_hz": 0,
    "step_size": 0
  },
  "REX": {
    "lock": 3545453316,
    "key": 0,
    "command": "none",
    "auth": 2210719614
  },
  "connected": true,
  "rssi": -72,
  "snooze": {
    "active": false,
    "startTime": ""
  },
  "nightlightColor": {
    "r": 0,
    "g": 0,
    "b": 0,
    "w": 65535
  }
}
```

This is the state after one left click on the device:

```json
{
  "env": "prod",
  "alarmsDisabled": false,
  "nightlightOn": false,
  "nightlightIntensity": 32780,
  "toddlerLockOn": false,
  "snoozeDuration": 540,
  "knockThreshold": 120,
  "knockDuration": 1,
  "activeTap": true,
  "knockAxis": 1,
  "current": {
    "srId": 156668225,
    "playing": "routine",
    "step": 1,
    "color": {
      "id": 602,
      "r": 65535,
      "g": 47031,
      "b": 0,
      "w": 8738,
      "i": 65535,
      "duration": 3600,
      "until": "duration"
    },
    "sound": {
      "id": 10150,
      "v": 39531,
      "mute": false,
      "url": "https://downloads.ctfassets.net/hlsdh3zwyrtx/2zTn8QBjnJwwnYhawYzrlD/14c0766acda9a06f113ffdff2495f878/RV4_Sleep_Brown_CGMasterV5_20221020.wav",
      "duration": 3600,
      "until": "duration"
    },
    "paused": false
  },
  "dataVersion": "20250823151851237",
  "sleepScene": {
    "srId": 0,
    "enabled": false
  },
  "timer": {
    "s": "2024-05-12 00:26:40",
    "d": 0
  },
  "streaming": {
    "status": "none"
  },
  "timezone": "America/New_York",
  "rF": {
    "v": "10.1.575",
    "i": true,
    "u": "https://firmware.hatchbaby.com/prod/restoreV4/restoreV4_DFU_10_1_575.json?f=g"
  },
  "deviceInfo": {
    "f": "10.1.575",
    "fR": 0,
    "hwVersion": "10.0.2",
    "powerStatus": 0,
    "sdCardVersionInfo": {
      "releaseDate": "unknown",
      "hashType": "md5",
      "hashCode": "0xe4e5d909ca46a1633e42ffe7b99e12ae"
    }
  },
  "clock": {
    "i": 429531702,
    "turnOffAt": "23:59:00",
    "turnOnAt": "07:00:00",
    "turnDimAt": "23:59:00",
    "turnBrightAt": "07:00:00",
    "flags": 34816,
    "turnOffMode": "custom"
  },
  "toddlerLock": {
    "turnOffAt": "00:00:00",
    "turnOnAt": "00:00:00",
    "turnOnMode": "never"
  },
  "lucky": 0,
  "LDR": "UNPOWERED",
  "LWTP": false,
  "debug": 0,
  "hwDebugFlags": 0,
  "logging": 37,
  "owned": true,
  "lastReset": "PowerOn",
  "ota": {
    "status": "none",
    "downloadProgress": 0,
    "installProgress": 0
  },
  "touch": {
    "flags": 0,
    "poll_rate_hz": 0,
    "refire_delay_ms": 0,
    "refire_rate_hz": 0,
    "step_size": 0
  },
  "REX": {
    "lock": 3545453316,
    "key": 0,
    "command": "none",
    "auth": 2210719614
  },
  "connected": true,
  "rssi": -75,
  "snooze": {
    "active": false,
    "startTime": ""
  },
  "nightlightColor": {
    "r": 0,
    "g": 0,
    "b": 0,
    "w": 65535
  }
}
```

**Comment by Tmatz27** [comment]
ok I changed my homebridge docker image and now I am pulling in the hatch as an accessory and seeing just the on/off switch! I have it ported over to home assistant and its all working!

**Comment by dgreif** [comment]
Thanks for testing it out. `6.0.0` has been released 🎉

**Comment by vdiaz311** [comment]
Awesome work guys. I now have a functioning on/off switch for my wife’s restore in home kit. Thanks for adding the feature!
